### PR TITLE
Submenu flipping with Popper.js!

### LINKED ISSUE #849: Allow submenus to be positioned
Currently if a menu is at the bottom of the screen and has a submenu that extends past the screen, you can never click on the submenu because it expands the screen size, closes, reopens, and bounces back and forth

If submenus were able to position to the top instead of the bottom this would not happen

Related video: https://pl.ntr/ep

### LINKED ISSUE #1640: MenuItem's submenu is closed when clicking the MenuItem
### Bug report

- Blueprintjs core v1.29.1
- Chrome v61

#### Steps to reproduce

1. Visit the v1.29.1 docs (this has been around for awhile though, i think) [Context Menu page](http://blueprintjs.com/docs/#core/components/context-menu)
1. Trigger the context menu & hover over a MenuItem with a submenu (e.g. 'Insert..."), the submenu will appear
1. Click the same MenuItem ('Insert...'). The submenu will disappear.
1. The only way to get the submenu back is to re-hover over the MenuItem

#### Actual behavior

Clicking on a MenuItem with a submenu will close the submenu, and subsequent clicks do not cause it to re-appear

#### Expected behavior

Most context menus I use, the submenu will persist regardless of # of clicks on the primary menu item.

Thanks for what you folks are doing with this library, it's great.

### LINKED ISSUE #2008: Fix Menu tests skip'd in the Popover => Popover2 transition
I `it.skip`'d the following tests during the 2.0 transition promoting `Popover2` to the core popover component. Need to get them working again and re-enable them:

```tsx
// menuTests.tsx
- "children will continue displaying in the same direction if possible"
- "children will flip direction after no more room in the existing direction"
```

Note: anecdotally, the functionality above _does_ still work in practice; it's just the tests that are failing.

### PR DESCRIPTION
**Pull request by giladgray**
#### Changes proposed in this pull request:

- 🔥  remove `MenuItem` `useSmartPositioning` and `submenuViewportMargin` props, and all associated logic, as `Popper.js` beautifully handles flipping when a menu reaches the edge.
  - also remove the tests cuz it's all in Popper now (#2008)
  - this functionality had _zero usages_ that I could find
  - Popper's logic can be overridden through `MenuItem` `popoverProps.modifiers`

#### Screenshot

notice how the menu never leaves the viewport (or causes it to scroll #849)!!

submenu-flipping

### CODE REVIEW

**Comment by llorca** [comment]
🔥 

Can you bring back the original alignment of submenus? It's off by 5px in both directions
screen shot 2018-01-29 at 8 37 00 pm

File: packages/core/src/components/menu/menu.md [hunk 1]
```diff
@@ -83,25 +83,34 @@ To add a submenu to a `Menu`, simply nest `MenuItem`s within another `MenuItem`.
 The submenu opens to the right of its parent by default, but will adjust and flip to the left if
 there is not enough room to the right.
 
+`Menu` provides two `submenu` props to adjust this flipping behavior: you can customize the boundary element
+and the padding within that boundary element. Note that a `MenuItem` _must be inside_ a `Menu` element
+for these props to affect the submenus. On standalone `MenuItem`s (without a parent `Menu`, an anti-pattern) you can
+use `popoverProps.modifiers` directly to achieve the same result.
```
Comment (adidahiya) [inline packages/core/src/components/menu/menu.md]: If it's an anti-pattern, why even mention it?

File: packages/core/src/components/menu/context.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+export interface IMenuItemContext {
+    getSubmenuPopperModifiers?(): Popper.Modifiers;
+}
+
+export const MenuItemContextTypes: React.ValidationMap<IMenuItemContext> = {
+    getSubmenuPopperModifiers: assertFunctionProp,
+};
+
+// simple alternative to prop-types dependency
```
[the remark this answers is not in the document]
Comment (giladgray) [reply packages/core/src/components/menu/context.ts]: well dang, i didn't realize react depended on it!

```
blueprint git:(gg/submenus) yarn why prop-types
yarn why v1.3.2
[1/4] 🤔  Why do we have the module "prop-types"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
[4/4] 🚡  Calculating file sizes...
info Reasons this module exists
   - "@blueprintjs/core#react-popper" depends on it
   - "@blueprintjs/core#react" depends on it
   - "@blueprintjs/core#react-transition-group" depends on it
```

[the remark this answers is not in the document]
Comment (giladgray) [reply packages/core/src/components/menu/context.ts]: no-implicit-dependency-wise, is `prop-types` a new peer dep?

File: packages/core/src/components/menu/menuItem.tsx [hunk 3]
```diff
@@ -136,141 +98,58 @@ export class MenuItem extends AbstractPureComponent<IMenuItemProps, IMenuItemSta
...
-        // this ensures that the left and right measurements represent a submenu opened to the right
-        let submenuLeft = submenuRect.left;
-        let submenuRight = submenuRect.right;
-        if (this.state.alignLeft) {
-            submenuLeft += adjustmentWidth;
-            submenuRight += adjustmentWidth;
-        }
+        // NOTE: use .pt-menu directly because using Menu would start a new context tree and
+        // we'd have to pass through the submenu props. this is just simpler.
+        const submenuContent = <ul className={Classes.MENU}>{this.renderChildren()}</ul>;
```
[the remark this answers is not in the document]
Comment (giladgray) [reply packages/core/src/components/menu/menuItem.tsx]: hmm yes, makes sense! surprising that never came up.

File: packages/core/src/components/menu/menuItem.tsx [hunk 4]
```diff
@@ -136,141 +98,58 @@ export class MenuItem extends AbstractPureComponent<IMenuItemProps, IMenuItemSta
...
+        const popoverModifiers = safeInvoke(this.context.getSubmenuPopperModifiers);
 
-        // this ensures that the left and right measurements represent a submenu opened to the right
-        let submenuLeft = submenuRect.left;
-        let submenuRight = submenuRect.right;
-        if (this.state.alignLeft) {
-            submenuLeft += adjustmentWidth;
-            submenuRight += adjustmentWidth;
-        }
+        // NOTE: use .pt-menu directly because using Menu would start a new context tree and
```
[the remark this answers is not in the document]
Comment (giladgray) [reply packages/core/src/components/menu/menuItem.tsx]: new context tree: the nested `Menu` overwrites the same `context` key with a function that returns _its_ default props, not the parent's. by not using `Menu`, the root context gets passed all the way down.

[the remark this answers is not in the document]
Comment (giladgray) [reply packages/core/src/components/menu/menuItem.tsx]: it's doable with `Menu`, i'll update the usage so you can see. don't need to do any constructor stuff, i can just pass parent props down.

File: packages/core/src/components/menu/menuItem.tsx [hunk 5]
```diff
@@ -136,141 +98,58 @@ export class MenuItem extends AbstractPureComponent<IMenuItemProps, IMenuItemSta
...
-            return;
-        }
-
-        const submenuRect = this.popoverElement.getBoundingClientRect();
-        const parentWidth = this.liElement.parentElement.getBoundingClientRect().width;
-        const adjustmentWidth = submenuRect.width + parentWidth;
+    private renderPopover(content: JSX.Element) {
+        const { disabled, popoverProps } = this.props;
+        const popoverClasses = classNames(Classes.MENU_SUBMENU, popoverProps.popoverClassName);
+        // getSubmenuPopperModifiers will not be defined if `MenuItem` used outside a `Menu`.
```
[the remark this answers is not in the document]
Comment (giladgray) [reply packages/core/src/components/menu/menuItem.tsx]: 🤔 i'm hesitant to throw errors in cases where it's not _absolutely_ fatal. pretty easy to handle the no-context case here.

File: packages/core/src/components/menu/context.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (adidahiya) [reply packages/core/src/components/menu/context.ts]: no, it's a regular dependency, not a peerDep.

File: packages/core/src/components/menu/menu.md [hunk 6]
```diff
@@ -83,25 +83,34 @@ To add a submenu to a `Menu`, simply nest `MenuItem`s within another `MenuItem`.
 The submenu opens to the right of its parent by default, but will adjust and flip to the left if
 there is not enough room to the right.
 
+`Menu` provides two `submenu` props to adjust this flipping behavior: you can customize the boundary element
+and the padding within that boundary element. Note that a `MenuItem` _must be inside_ a `Menu` element
```
[the remark this answers is not in the document]
Comment (giladgray) [reply packages/core/src/components/menu/menu.md]: you might want to keep your `Menu` inside a container other than the viewport, maybe on a `canvas`. padding i think is less useful and less likely to be used so i could be convinced to remove it, but it's easy enough to support.

File: packages/core/src/components/menu/menu.md (hunk 1 above)
Comment (giladgray) [reply packages/core/src/components/menu/menu.md]: yeah i had this exact thought. i figure it's going to happen (is already happening) so why not pass judgment on it.

**Comment by adidahiya** [comment]
- We should remove `submenuBoundaryPadding ` as a prop. Please don't add it just because it's "easy enough to support" -- IMO nothing ever is. We can add it back later if users ask for it.
- We should also strongly consider removing `submenuBoundaryElement` until users ask for it. Can you demonstrate a legit use case for that here in this PR (ideally in the docs example)?

File: packages/core/src/components/menu/context.ts [hunk 7]
```diff
@@ -4,21 +4,12 @@
  * Licensed under the terms of the LICENSE file distributed with this project.
  */
 
-import { isFunction } from "../../common/utils";
+import { func } from "prop-types";
```
Comment (adidahiya) [inline packages/core/src/components/menu/context.ts]: not a fan of this import, the symbols are too terse. for example another possible import is `any`, which would be very confusing to import. please switch to `import * as PropTypes`

File: packages/docs-theme/src/common/context.ts [hunk 8]
```diff
@@ -4,14 +4,14 @@
...
 
-import { Utils } from "@blueprintjs/core";
 import {
     IBlock,
     IKssPluginData,
     IMarkdownPluginData,
     ITsDocBase,
     ITypescriptPluginData,
 } from "documentalist/dist/client";
+import { func } from "prop-types";
```
Comment (adidahiya) [inline packages/docs-theme/src/common/context.ts]: same here, use namespace import

File: packages/core/src/components/menu/menuItem.tsx [hunk 9]
```diff
@@ -63,11 +63,11 @@ export class MenuItem extends AbstractPureComponent<IMenuItemProps> {
...
 
     public render() {
-        const { children, disabled, label, submenu } = this.props;
-        const hasSubmenu = children != null || submenu != null;
-        const liClasses = classNames({
-            [Classes.MENU_SUBMENU]: hasSubmenu,
-        });
+        const { disabled, label } = this.props;
+        const submenuChildren = this.renderSubmenuChildren();
+        const hasSubmenu = submenuChildren != null;
```
Comment (adidahiya) [inline packages/core/src/components/menu/menuItem.tsx]: nit: only used once, can be inlined

File: packages/core/src/components/menu/context.ts (hunk 7 above)
Comment (giladgray) [reply packages/core/src/components/menu/context.ts]: yeah i felt the terseness. went with this for theoretical tree-shaking purposes, but i agree the `* as` import makes a lot more sense here.

**Comment by giladgray** [comment]
@adidahiya ok, i'm onboard to remove the two props. the only real benefit they conferred was the cascading behavior to all nested submenus. the behavior is still entirely possible through `popoverProps.modifier` on `MenuItem`.

File: packages/core/src/components/menu/menuItem.tsx (hunk 9 above)
Comment (giladgray) [reply packages/core/src/components/menu/menuItem.tsx]: used twice, or i would have already inlined it 😄  (other usage is hidden in collapsed block, line 66)

**Comment by giladgray** [comment]
#2058 should fix test failures here.

File: packages/core/src/components/menu/menuItem.tsx [hunk 10]
```diff
@@ -25,12 +25,11 @@ export interface IMenuItemProps extends IActionProps, ILinkProps {
      */
     label?: string | JSX.Element;
 
-    /** Props to spread to `Popover`. Note that `content` cannot be changed. */
-    popoverProps?: Partial<IPopoverProps> & object;
+    /** Props to spread to `Popover`. Note that `content` and `minimal` cannot be changed. */
+    popoverProps?: Partial<IPopoverProps>;
```
Comment (adidahiya) [inline packages/core/src/components/menu/menuItem.tsx]: `& object` was there intentionally... IIRC without it, you can assign literally anything non-null to this field.

Comment (giladgray) [reply packages/core/src/components/menu/menuItem.tsx]: `& object` messes up the docs, which i take to be more important. a prop called `*Props` is expected to be an object.
screen shot 2018-01-31 at 3 06 45 pm

**Pull request merged**